### PR TITLE
Add incoming pod metrics to scheduler queue.

### LINKED ISSUE #83410: Add a metric to tracking incoming pods into scheduler queues
**What would you like to be added**:

metric name | metric definition | Description
-- | -- | --
scheduler_queue_incoming_pods_total | Type: CounterLabels:event={see below},queue={“active”, “backoff”, “unschedulable”} | Number of pods added to scheduling queues.

**Event**
There are a bunch of events happening that causes pods to be added to scheduling queues.
* "PodAdd"
Scheduler watches API server for pod creation and adds new pod to its active queue.
* "ScheduleAttemptFailure"
When a schedule attempt fails, we add the pod back to either backoff queue or unschedulable queue.
* "BackoffComplete"
When a pod in the backoff queue finishes backoff, it's added back to the active queue.
* "UnschedulableTimeout"
When a pod stays in the unschedulable queue for more than a minute, it's added back to the active queue.
* “PvAdd”, “PvUpdate”, “PvcAdd”, “PvcUpdate”, “StorageClassAdd”, “ServiceAdd”, “ServiceUpdate”, “ServiceDelete”, “CSINodeAdd”, “CSINodeUpdate”, “CSINodeDelete”
These are cluster events that may cause an unschedulable pod to be schedulable again. When these happen, we flush the unschedulable pods to either active queue or backoff queue.
* "PodMatchingAffinityScheduled"
Once a pod with matching affinity term is scheduled, related pods may become schedulable again. When this happens, we flush the unschedulable pods to either active queue or backoff queue.

**Why is this needed**:
This gives us more insights into the dynamics of the scheduling queue. 

/sig scheduling

### PR DESCRIPTION
**Pull request by liu-cong**
**What type of PR is this?**

/kind feature
/kind cleanup

**What this PR does / why we need it**:
#83410

**Which issue(s) this PR fixes**:

Fixes #83410

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Add incoming pods metrics to scheduler queue.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

/sig scheduling

### CODE REVIEW

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 1]
```diff
@@ -58,7 +58,6 @@ const unschedulableQTimeInterval = 60 * time.Second
 // makes it easy to use those data structures as a SchedulingQueue.
 type SchedulingQueue interface {
 	Add(pod *v1.Pod) error
-	AddIfNotPresent(pod *v1.Pod) error
```
Comment (liu-cong) [inline pkg/scheduler/internal/queue/scheduling_queue.go]: I removed this since it's not called anywhere.

File: pkg/scheduler/eventhandlers.go [hunk 2]
```diff
@@ -122,11 +124,11 @@ func (sched *Scheduler) updateNodeInCache(oldObj, newObj interface{}) {
 
 	// Only activate unschedulable pods if the node became more schedulable.
 	// We skip the node property comparison when there is no unschedulable pods in the queue
-	// to save processing cycles. We still trigger a move to active queue to cover the case
-	// that a pod being processed by the scheduler is determined unschedulable. We want this
-	// pod to be reevaluated when a change in the cluster happens.
-	if sched.SchedulingQueue.NumUnschedulablePods() == 0 || nodeSchedulingPropertiesChanged(newNode, oldNode) {
-		sched.SchedulingQueue.MoveAllToActiveQueue()
+	// to save processing cycles.
+	if sched.SchedulingQueue.NumUnschedulablePods() != 0 {
```
Comment (liu-cong) [inline pkg/scheduler/eventhandlers.go]: Please pay attention to this change.
The previous comment/code doesn't make sense to me. If `NumUnschedulablePods` is 0, when you call `MoveAllToActiveQueue`, it scans the `unschedulableQ` map, which should be empty.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 3]
```diff
@@ -502,68 +481,53 @@ func (p *PriorityQueue) Delete(pod *v1.Pod) error {
...
+	p.movePodsToActiveOrBackoffQueue(unschedulablePods, event)
 	p.moveRequestCycle = p.schedulingCycle
 	p.cond.Broadcast()
 }
 
 // NOTE: this function assumes lock has been acquired in caller
-func (p *PriorityQueue) movePodsToActiveQueue(podInfoList []*framework.PodInfo) {
+func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(podInfoList []*framework.PodInfo, event string) {
+	var newBackOff, newActive float64
+	defer metrics.SchedulerQueueIncomingPods.WithLabelValues("backoff", event).Add(newBackOff)
```
Comment (wgliang) [inline pkg/scheduler/internal/queue/scheduling_queue.go]: I would recommend merging into a defer function.

**Review by wgliang — CHANGES_REQUESTED** [review]
Others LGTM

**Review by ahg-g — COMMENTED** [review]
looks good overall, a couple of comments just to be on the safe side.

**Comment by liu-cong** [comment]
Thanks for the review! 
I am going to  add some unit tests and address the comments.

**Comment by ahg-g** [comment]
can you please rebase

**Comment by liu-cong** [comment]
> can you please rebase

Rebased.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 4]
```diff
@@ -550,68 +530,56 @@ func (p *PriorityQueue) Delete(pod *v1.Pod) error {
...
 
 // NOTE: this function assumes lock has been acquired in caller
-func (p *PriorityQueue) movePodsToActiveQueue(podInfoList []*framework.PodInfo) {
+func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(podInfoList []*framework.PodInfo, event string) {
+	var newBackOff, newActive float64
+	defer func() {
+		metrics.SchedulerQueueIncomingPods.WithLabelValues("backoff", event).Add(newBackOff)
+		metrics.SchedulerQueueIncomingPods.WithLabelValues("active", event).Add(newActive)
+
+	}()
```
[the remark this answers is not in the document]
Comment (liu-cong) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: No particular reason, I think it's just easier to follow: you define the `newBackOff, newBackOff` and then **"immediately"** record them in a `defer` statement. 

Refactored to not use `defer`

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 5]
```diff
@@ -390,6 +366,8 @@ func (p *PriorityQueue) flushBackoffQCompleted() {
 	p.lock.Lock()
 	defer p.lock.Unlock()
 
+	var newActive float64
+	defer func() { metrics.SchedulerQueueIncomingPods.WithLabelValues("active", BackoffComplete).Add(newActive) }()
```
[the remark this answers is not in the document]
Comment (liu-cong) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: Refactored to not use defer

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 6]
```diff
@@ -550,68 +530,56 @@ func (p *PriorityQueue) Delete(pod *v1.Pod) error {
...
+func (p *PriorityQueue) MoveAllToActiveOrBackoffQueue(event string) {
 	p.lock.Lock()
 	defer p.lock.Unlock()
-
-	// There is a chance of errors when adding pods to other queues,
-	// we make a temporary slice to store the pods,
-	// since the probability is low, we set its len to 0
-	addErrorPods := make([]*framework.PodInfo, 0)
-
+	var unschedulablePods []*framework.PodInfo
```
[the remark this answers is not in the document]
Comment (liu-cong) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: Good catch. Refactored.

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (liu-cong) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: Actually, we need a `defer` here. In the `for` loop, there are multiple cases where we end the loop and return. Without a `defer`, we need to catch all those branches and record the metric properly.

[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: how is that different for ```newActive++```? why not replace newActive++ with Inc() in both places?

[the remark this answers is not in the document]
Comment (liu-cong) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: Sorry didn't realize there is already an `Inc` method. Yes calling `Inc()` is better.
Thanks!

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 7]
```diff
@@ -550,68 +527,52 @@ func (p *PriorityQueue) Delete(pod *v1.Pod) error {
...
-				addErrorPods = append(addErrorPods, pInfo)
-			}
-		}
-	}
-	p.unschedulableQ.clear()
-	// Adding pods that we could not move to Active queue or Backoff queue back to the Unschedulable queue
-	for _, podInfo := range addErrorPods {
-		p.unschedulableQ.addOrUpdate(podInfo)
+		unschedulablePods[i] = pInfo
+		i++
```
Comment (ahg-g) [inline pkg/scheduler/internal/queue/scheduling_queue.go]: you can still use append here.

**Review by ahg-g — COMMENTED** [review]
/approve

let me know if you want to change that nit, otherwise I will lgtm

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 7 above)
Comment (liu-cong) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: I think `append` will increase the size. See https://play.golang.org/p/vGx-sTuV9Bg

Comment (ahg-g) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: not if you do this: https://play.golang.org/p/F-78XuIT83x

Comment (liu-cong) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: Done.Thanks.

**Pull request merged**